[fastlane] fix keychain shellescaping

datajockey opened the pull request:
<!-- Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

There are issues with keychains being escaped and/or used in an inconsistent way. 
See issue #14310 for a description and reproduction steps.

### Description
<!-- Describe your changes in detail -->
<!-- Please describe in detail how you tested your changes. -->

1. The first commit ensures that keychain creation security commands use the escaped keychain path and avoids double escaping via `.shelljoin` 

   From the shelljoin docs:
   > All elements are joined into a single string with fields separated by a space, where each element is escaped for the Bourne shell and stringified using to_s.

2. The second commit removes the quotes around the escaped keychain path while importing.

   From the shellescape docs:
   > Note that a resulted string should be used unquoted and is not intended for use in double quotes nor in single quotes.

3. ~~The third commit constrains the certificate check to the provided keychain. 
   Through my test Fastfile (see #14310), I discovered that if my credentials were in any keychain in the search path, they would not be installed in the provided/target keychain. 
   I thought this was odd so I added this commit even though it is not directly related to the underlying issues.~~ 
   **Upate:** This commit was moved to a new branch to be evaluated a separate PR.

Testing consisted of manually running a Fastfile (see #14310) with control and test scenarios and comparison with command line executions of the security command. 

The unit tests do not actually execute the commands, but compare the command that would be issued to an expected value. I updated the unit tests to reflect my changes and identify things I missed.

**Note:** I doubt it is a complete solution but a step in the right direction. Please let me know of any changes or extensions. I hope this helps!

datajockey committed:
The shellescaped keychain_path should be unquoted.

datajockey committed:
Use escaped_keychain_path with 'security' commands for keychain manipulation.
Watchout for double escaping with .shelljoin.

datajockey commented:
I signed it!

datajockey commented:
The third commit to check if the certs were installed in the target keychain is problematic. 
I can drop this commit or with some direction, I can fix it up. 
Please advise.

janpio commented:
Without looking at it in detail, best move the third commit to its own PR so we can look at this different thing in isolation. (Thanks for the great separation into commits etc - makes it easy to follow what is going on here.)

janpio commented on fastlane/spec/actions_specs/import_certificate_spec.rb:
```diff
@@ -11,7 +11,7 @@
 
         keychain_path = File.expand_path(File.join('~', 'Library', 'Keychains', keychain))
-        expected_command = "security import #{cert_name} -k '#{keychain_path}' -P #{password} -T /usr/bin/codesign -T /usr/bin/security 1> /dev/null"
+        expected_command = "security import #{cert_name} -k #{keychain_path} -P #{password} -T /usr/bin/codesign -T /usr/bin/security 1> /dev/null"
```
isn't this missing `.shellescape` then?

datajockey replied:
Escaping is a no-op here because there are no special characters in the test parameters. 
Good catch and the other parameters should be escaped too.

janpio commented on fastlane/spec/actions_specs/import_certificate_spec.rb:
```diff
@@ -66,7 +66,7 @@
 
         keychain_path = File.expand_path(File.join('~', 'Library', 'Keychains', keychain))
-        expected_command = "security import #{cert_name} -k '#{keychain_path}' -P #{password} -T /usr/bin/codesign -T /usr/bin/security"
+        expected_command = "security import #{cert_name} -k #{keychain_path} -P #{password} -T /usr/bin/codesign -T /usr/bin/security"
```
Same here:
isn't this missing `.shellescape` then?

datajockey replied:
Same here:
> Escaping is a no-op here because there are no special characters in the test parameters. 
> Good catch and the other parameters should be escaped too.

datajockey commented:
I moved the third commit to a new branch as recommended and reset HEAD to second commit.  
I addressed the missing `.shellescape` comments in the amended second commit and force pushed the branch.

janpio commented:
Hm yeah, the noop in the specs are strange to decide - you don't need it, but it makes it more correct :/ Thanks for changing.

janpio commented on match/spec/utils_spec.rb:
```diff
@@ -6,7 +6,7 @@
     describe 'import' do
       it 'finds a normal keychain name relative to ~/Library/Keychains' do
-        expected_command = "security import item.path -k '#{Dir.home}/Library/Keychains/login.keychain' -P #{''.shellescape} -T /usr/bin/codesign -T /usr/bin/security 1> /dev/null"
+        expected_command = "security import item.path -k #{Dir.home}/Library/Keychains/login.keychain -P #{''.shellescape} -T /usr/bin/codesign -T /usr/bin/security 1> /dev/null"
```
Isn't this missing a `.shellescaoe` for the `-k` value here? (relevant e.g. when `Dir.home` has a space)

datajockey replied:
@janpio, Yes, I noticed other cases where the `spec` files needed to be updated with `.shellescape` after I pushed my initial update. 

I have not been unable to get back to them, but I will address these today.

datajockey replied:
I pushed an update to these.

janpio commented on match/spec/utils_spec.rb:
```diff
@@ -25,7 +25,7 @@
         tmp_path = Dir.mktmpdir
         keychain = "#{tmp_path}/my/special.keychain"
-        expected_command = "security import item.path -k '#{keychain}' -P #{''.shellescape} -T /usr/bin/codesign -T /usr/bin/security 1> /dev/null"
+        expected_command = "security import item.path -k #{keychain} -P #{''.shellescape} -T /usr/bin/codesign -T /usr/bin/security 1> /dev/null"
```
same as above

datajockey replied:
Same here.

janpio commented on match/spec/utils_spec.rb:
```diff
@@ -50,7 +50,7 @@
 
       it "tries to find the macOS Sierra keychain too" do
-        expected_command = "security import item.path -k '#{Dir.home}/Library/Keychains/login.keychain-db' -P #{''.shellescape} -T /usr/bin/codesign -T /usr/bin/security 1> /dev/null"
+        expected_command = "security import item.path -k #{Dir.home}/Library/Keychains/login.keychain-db -P #{''.shellescape} -T /usr/bin/codesign -T /usr/bin/security 1> /dev/null"
```
and again

datajockey replied:
Same here.

janpio commented on fastlane/lib/fastlane/actions/create_keychain.rb:
```diff
@@ -38,34 +38,35 @@ def self.run(params)
 
         if params[:add_to_search_list]
           keychains = Action.sh("security list-keychains -d user").shellsplit
+          # Do not use 'escaped_keychain_path' here because .shelljoin will cause double escaping
```
all the changes in this file are just so this is possible, correct?

datajockey replied:
I did this to be more explicit with what is being used throughout the file.  I.e. `escaped_keychain_path` vs sprinkling `.shellescape` in all of these cases except for the `.shelljoin` case where we want to use the plain `keychain_path`.

datajockey replied:
I did this to be more explicit with what is being used throughout the file.  I.e. `escaped_keychain_path` vs sprinkling `.shellescape` in all of these cases except for the `.shelljoin` case where we want to use the plain `keychain_path`.

datajockey commented on fastlane/spec/actions_specs/create_keychain_spec.rb:
```diff
@@ -2,106 +2,122 @@
+        expect(result[1]).to include(keychain_path.to_s)
         expect(result[2]).to start_with("security list-keychains -s")
-        expect(result[2]).to end_with(File.expand_path('~/Library/Keychains/test.keychain').to_s)
+        expect(result[2]).to end_with(File.expand_path(keychain_path.to_s).to_s)
```
@janpio Note that I did not change this to `keychain_path.shellescape` because of the link in the link causes the keychain path to be expanded:
https://github.com/fastlane/fastlane/pull/14311/files#diff-0231a345cb1ba3f328959032b555024cR57

So `~/Library/Keychains/...` becomes `/Users/foo/Library/Keychains/...` in this test case. It seems a little odd so if you have any thoughts, let me know.

datajockey replied:
@janpio Nevermind the note above, I realized my mistake and resolved the use of `.shellescape` in this test case.

joshdholtz reviewed (approved):
@datajockey This PR looks good! Thanks for contribution here 👍 And thanks for all the reviewing, @janpio ❤️

datajockey commented:
@joshdholtz Glad I can help out! 
@janpio Thank you for your help and guidance.

Following up on the problematic third commit which @janpio suggested that I break out into a separate PR. It was meant to address the discovery that signing credentials were not being installed in the target keychain if they were already installed in another keychain on the search path. This non-deterministic behavior seemed odd to me, but maybe a desired behavior so I wanted to get your thoughts. 

My initial commit for this issue passed all local rspec tests, but failed on the Ubuntu and Appveyor checks. It was unclear to me what conditions were being tested that lead to the failures, so there is work to do to isolate and ready a PR.

janpio commented:
Feel free to create a `[WIP]` PR @datajockey, then we can maybe help debug this further. (And also to avoid this getting forgotten and lost)

datajockey commented:
Will do @janpio 
👍
